### ISSUE TITLE
Feature request: add import/require information in plugin OnResolveArgs

### ISSUE THREAD

**Issue by yyx990803** [evanw/esbuild#879]
Currently in the `onResolve` hook of a plugin, we don't have information on whether the request was made via ESM import or CJS require. This information, however, is somewhat important when resolving Node dependencies, especially those with explicit [sub path `exports` definitions](https://nodejs.org/api/packages.html#packages_subpath_exports).

Take [@babel/runtime](https://unpkg.com/browse/@babel/runtime@7.13.7/package.json) as an example: it defines exports for all its helpers, e.g.

```json
"./helpers/interopRequireDefault": [
      {
        "node": "./helpers/interopRequireDefault/index.js",
        "require": "./helpers/interopRequireDefault/index.js",
        "default": "./helpers/interopRequireDefault/_index.mjs"
      },
      "./helpers/interopRequireDefault/index.js"
    ],
```

It probably does this because it wants to allow users to do both

```js
// gets the cjs version
const foo = require('@babel/runtime/helpers/foo')

// resolves the mjs version
import foo from '@babel/runtime/helpers/foo'
```

Without the esm vs. cjs information though, the plugin `onResolve` hook can only guess which one it should resolve to. In the case for Vite, it will prefer ESM entries, so it resolves to the mjs version which, when required, fails to run because the helper is the default export and needs to be accessed as `foo.default`.

Since esbuild seems to allow the mix of ESM and CJS, it would be beneficial to somehow expose that information to plugins sp that resolvers can properly resolve conditional exports.

**Comment by evanw** [comment]
Makes sense. This is already available internally so exposing this is straightforward. Here's what will be in the next release (from the release notes):

> * Provide `kind` to `onResolve` plugins ([#879](https://github.com/evanw/esbuild/issues/879))
> 
>     Plugins that add `onResolve` callbacks now have access to the `kind` parameter which tells you what kind of import is being resolved. It will be one of the following values:
> 
>     * `"entry-point"` in JS (`api.ResolveEntryPoint` in Go)
> 
>         An entry point provided by the user
> 
>     * `"import-statement"` in JS (`api.ResolveJSImportStatement` in Go)
> 
>         A JavaScript `import` or `export` statement
> 
>     * `"require-call"` in JS (`api.ResolveJSRequireCall` in Go)
> 
>         A JavaScript call to `require(...)` with a string argument
> 
>     * `"dynamic-import"` in JS (`api.ResolveJSDynamicImport` in Go)
> 
>         A JavaScript `import(...)` expression with a string argument
> 
>     * `"require-resolve"` in JS (`api.ResolveJSRequireResolve` in Go)
> 
>         A JavaScript call to `require.resolve(...)` with a string argument
> 
>     * `"import-rule"` in JS (`api.ResolveCSSImportRule` in Go)
> 
>         A CSS `@import` rule
> 
>     * `"url-token"` in JS (`api.ResolveCSSURLToken` in Go)
> 
>         A CSS `url(...)` token
> 
>     These values are pretty much identical to the `kind` field in the JSON metadata file.

**Comment by evanw** [comment]
FYI this was just published as version 0.8.52.

**Comment by yyx990803** [comment]
Thanks!
